Terminal.Gui/Application/Application.Mouse.cs

```diff
@@ -249,6 +249,15 @@ internal static void RaiseMouseEvent (MouseEventArgs mouseEvent)
                 View = deepestViewUnderMouse
             };
         }
+
+        // TODO: There should be a clearner way to to this
+        // Add a timeout to ensure at least one application iteration happens after each
+        // mouse event, enabling Refresh to be called.
+        Application.AddTimeout(new (0,0,0,0, 200),
```

Comment:
I really don't understand the necessity of this call here. Any view that will handle the mouse is responsible to call `SetNeedsLayout` to recalculate Pos/Dim and `SetNeedsDisplay` or `SetSubViewNeedsDisplay` if they only needed to redraw. Calling `SetNeedsLayout` automatically force the necessity of calling `SetNeedsDisplay`. Calling `Refresh` in this `AddTimeout` will decrease the performance because will call `Refresh` more times than the necessary. Not all the mouse event need to call `Refresh` at all, unless the view needs to design effects on mouse move, where the view is responsible to set the above proprieties if needed. I think calling `Refresh` through the `RunIteration` is enough.

Reply:
Here's the problem that needs to be solved, that this works around:

In this PR, `Daw` called only from `Refresh` which is ONLY called at the end of an iteration. (There are some exceptions where some code still calls `View.Draw` or `Applicaiton.Refresh` directly, but those are either bugs or very special cases).

`Refresh` does this:

```cs
...

        Stack<Toplevel> redrawStack = new (TopLevels);

        while  (redrawStack.Count > 0)
        {
            Toplevel? tlToDraw = redrawStack.Pop ();

            if (clear || forceRedraw)
            {
                tlToDraw.SetNeedsDisplay ();
            }

            if (!(!View.CanBeVisible (tlToDraw) || tlToDraw is { NeedsDisplay: false, SubViewNeedsDisplay: false }))
            {
                tlToDraw.Draw ();

                if (redrawStack.TryPeek (out var nexToplevel))
                {
                    nexToplevel.SetNeedsDisplay();
                }
            }
        }

        Driver?.Refresh ();
```

This logic is currently not complete, because I was still debugging something, but it essentially ensures the top-most Toplevel draws last.

The problem is code like this:

```cs
Toplevel app = new () { "app"};

Button btn = new () { Title = "_Open Dialog" );
app.Add(btn);
btn.Accepting += (s,e) => { Application.Run (dlg); };

Application.Run(app);
```

If `btn` is pressed with the hotkey (`O`) everything is fine.  

- User presses `O` down
- `Application.RaiseKeydown` is called by the driver from `MainLoop.RunIteration ()`
  - Causes `btn.Accepting` -> `Application.Run(dlg)` 
  - `Run` calls `Begin` changes focus such that `btn` no longer has focus so it and `app` get `NeedsDisplay = true`.
  - `Run` calls `RunLoop` which BLOCKS waiting for the driver to generate an event
- User releases `O` key causing the driver to unblock
- `Application.RaiseKeydown` is called by the driver from `MainLoop.RunIteration ()`
  - The iteration calls `Refresh` and draws everything correctly.

Yay.

HOWEVER, If `btn` is pressed with the MOUSE:

- User presses `btn`
- `Application.RaiseMouseEvent` is called with `Button1Pressed` by the driver from `MainLoop.RunIteration ()`
- `Application.RaiseMouseEvent` is called with `Button1Released` by the driver from `MainLoop.RunIteration ()`
- `Application.RaiseMouseEvent` is called with `Button1Clicked` by the driver from `MainLoop.RunIteration ()`
  - Causes `btn.Accepting` -> `Application.Run(dlg)` 
  - `Run` calls `Begin` changes focus such that `btn` no longer has focus so it and `app` get `NeedsDisplay = true`.
  - `Run` calls `RunLoop` which BLOCKS waiting for the driver to generate an event

The Dialog is not rendered because we're blocked waiting for a new event from the driver. 

Move the mouse and `Refresh` gets called and all is good.
  
This is really hard to debug because `WindowsDriver` generates events on focus changes and each time you step in the debugger, it re-activates the terminal generating iterations. 

Anyway, here's what we don't want to do to fix this:

- Cause Iterations to happen all the time. 
- Call `Refresh` synchronously from `Run/Begin` (see rule above about only doing draw async).

The best solution I could find was to generate a Timer event after each mouse event.

I'm very much open to other ideas.

> Calling Refresh in this AddTimeout will decrease the performance because will call Refresh more times than the necessary.

This is no longer true with this PR because Refresh is a NOP if there's no `NeedsDisplay = true` anywhere.

Reply:
I already saw that a view that call SetFocus before call Application.Run loose the focus in the Add method. I don't know why.

Reply:
> The Dialog is not rendered because we're blocked waiting for a new event from the driver.
> 
When a new modal toplevel is loaded the `Begin` method must call the `Refresh` like it also call the `PositionCursor`. This is mandatory.
```cs
        toplevel.OnLoaded ();

        Refresh ();

        if (PositionCursor ())
        {
            Driver?.UpdateCursor ();
        }

```

> Move the mouse and `Refresh` gets called and all is good.
> 
If you call the `Refresh` as above move the mouse isn't necessary.

> This is really hard to debug because `WindowsDriver` generates events on focus changes and each time you step in the debugger, it re-activates the terminal generating iterations.
> 
Use `NetDriver` or `CursesDriver` to debug it because they don't generate events on focus changes.

> Anyway, here's what we don't want to do to fix this:
> 
> * Cause Iterations to happen all the time.

It's preferable to execute all the timers and idle handlers than having the drawing occurring all the time without need, which I already saw it's now a NOP which is very useful.

> * Call `Refresh` synchronously from `Run/Begin` (see rule above about only doing draw async).
> 
I don't agree not calling the `Refresh` to draw in the `Begin` method. If the rule must be imposed then the PositionCursor method mustn't also be called. The problem you describe that it's needed to move the mouse when running a `Dialog` also occurs when you start running the UICatalog because Begin doesn't call `Refresh`.

> The best solution I could find was to generate a Timer event after each mouse event.
> 
Not needed in a regular mouse event. If the Button only respond to the click event then the pressed and release only run `MainLoop.RunIteration ()` if there are timers or idle handlers to be executed and drawing if it's needed, like the button effects.

> I'm very much open to other ideas.
> 
I'm tying to give some.

> > Calling Refresh in this AddTimeout will decrease the performance because will call Refresh more times than the necessary.
> 
> This is no longer true with this PR because Refresh is a NOP if there's no `NeedsDisplay = true` anywhere.

I know and I like the solution added. It's the best place to have this control instead of in the `MainLoop.RunIteration`. But I think the `AddTimeout` in the `RaiseMouseEvent` is unnecessary.

Reply:
> When a new modal toplevel is loaded the Begin method must call the Refresh. This is mandatory.

The fact you say this means you are not understanding what I've done in this PR. 

Please dive deeper and see the purity of having Refresh only happen at the end of an iteration is the way to go.

Reply:
Actually, I think I just found a better solution that's more generic.

Instead of doing this in the mouse handler, add an Idle event at the end of `Application.Begin`:

```cs
        // Force an Idle event so that an Iteration (and Refresh) happen.
        Application.Invoke(() => {});
```

Reply:
I now understand, sorry. I like more this approach because it don't impose any timeout and wakeup immediately the mainloop. But you also can remove the call t the `PositionCursor` in the `Begin` method because it will be called in the `RunIteration`.
Another doubt I have is when a view only call the `SetNeedsDisplay` and not the `SetNeedsLayout` will the screen been drawing?

Reply:
Re: cursor - I'm intentionally ignoring working on that until we get to the PR for the issue on it (mobile so can't look up issue #). 

Re: when to call SetNeedsDisplay vs SetNeedsLayout. Not sure about that yet. Been thinking a lot about it though. Would love suggestions.

Reply:
If a view only need to redraw because something changed on a fixed size (all area or on a area part) then only call `SetNeedsDisplay` is enough. If a view need to redraw because something changed the position or the size then it's needed to call `SetNeedsLayout` which will impose a mandatory `SetNeedsDisplay` call.

Reply:
> If a view need to redraw because something changed the position or the size then it's needed to call `SetNeedsLayout` which will impose a mandatory `SetNeedsDisplay` call.

I forgot to mention that in this case the views under the current view must be flag to being redraw as well, but of course you already know.